File: 1605-minimum-number-of-days-to-make-m-bouquets/solution.java
class Solution {
    public int minDays(int[] bloomDay, int m, int k) {
        int low = 1;
        int n = bloomDay.length;
        int maxVal = Integer.MIN_VALUE;
        for(int i=0; i<n; i++) {
            maxVal = Math.max(maxVal, bloomDay[i]);
        }
        int high = maxVal;
        int ans = -1;

        if(m*k > n) return -1;

        while(low <= high) {
            int mid = low+(high-low)/2;
            if(canMakeBoquets(bloomDay, mid, m, k)) {
                ans = mid;
                high = mid-1;
            } else {
                low = mid+1;
            }
        }
        return ans;
    }

    public boolean canMakeBoquets(int[] bloomDay, int val, int m, int k) {
        int n = bloomDay.length;
        int numOfBoq = 0;
        int flower = 0;
        for(int i=0; i<n; i++) {
            // here val => currDay
            if(bloomDay[i] <= val) {
                // how many flowers are usable
                flower++;

                if(flower == k) {
                    numOfBoq++;
                    flower = 0;
                    if(numOfBoq == m) {
                        return true;
                    } 
                }
            } else {
                // flowers are not usable, so start from beginning
                flower = 0;
            }
        }
        return false; 
    }
}

File: 1605-minimum-number-of-days-to-make-m-bouquets/solution.swift
class Solution {
    func minDays(_ bloomDay: [Int], _ m: Int, _ k: Int) -> Int {
        var low = bloomDay.min()!
        var high = bloomDay.max()!

        if bloomDay.count < (m*k) {
            return -1
        }
        while low < high {
            var mid = low+(high-low)/2
            if canMake(mid) {
                high = mid
            } else {
                low = mid+1
            }
        }

        // i need to say if i can make m bouquets or not
        // to say if i can make, i need to check m = k adj flowers and no of days <= bloom day
        // Need this
        // flowers = consecutive flowers available
        // bouquets = bouquets formed
        func canMake(_ day: Int) -> Bool {
            var flowers = 0
            var m = m
            for bloom in bloomDay {
                if day >= bloom {
                    // make m bouquets
                    flowers += 1
                } else {
                    // break adjacency
                    flowers = 0
                }
                // 1 bouquet
                if flowers == k {
                    m -= 1
                    // reset , after making 1 bouquet
                    flowers = 0

                    if m == 0 {
                        return true
                    }
                }
            }
            return false
        }
        return high
    }
}
